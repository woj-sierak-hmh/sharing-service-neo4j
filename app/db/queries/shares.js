import session from '../connection.js';

export const createShare = async ({
  tenantRefId,
  sharerRefId,
  assetType,
  assetRefId,
  recipients,
}) => {
  const result = await session.run(
    'MERGE (org:Organization {orgRefId: {tenantRefId}}) ' +
      'MERGE (asset:Asset {assetRefId: {assetRefId}, assetType: {assetType}, isActive: "true"}) ' +
      'MERGE (org)-[:MASTER_OF]->(asset) ' +
      'MERGE (creator:User {userRefId: {sharerRefId}}) ' +
      'MERGE (creator)-[:CREATOR_OF]->(asset) ' +
      'FOREACH (ref IN {recipients} | ' +
      '  MERGE (recipient:User {userRefId: ref}) ' +
      '  MERGE (asset)-[:SHARED_WITH {type:"READ", createdDate: $createdDate}]->(recipient))',
    {
      tenantRefId,
      sharerRefId,
      assetType,
      assetRefId,
      createdDate: new Date().toISOString(),
      // temporarily...
      // recipient: `teacher` + Math.ceil(Math.random() * 100)
      recipients,
    }
  );
  session.close();
  return result;
};

export const getShares = async ({ tenantRefId, userRefId, assetType }) => {
  const results = await session.run(
    'MATCH (:User {userRefId: {userRefId}})<-[share:SHARED_WITH]-(asset:Asset {assetType: {assetType}})<-[:MASTER_OF]-(:Organization {orgRefId: {tenantRefId}}) ' +
      'WITH share, asset ' +
      'MATCH (asset)<-[:CREATOR_OF]-(owner:User) ' +
      'RETURN asset.assetRefId AS assetRefId, asset.assetType AS assetType, ' +
      'share.createdDate AS shareDate, owner.userRefId AS sharerRefId ' +
      'ORDER BY shareDate DESC',
    {
      tenantRefId,
      userRefId,
      assetType,
    }
  );
  const assets = results.records.map(r => {
    return {
      assetRefId: r.get('assetRefId'),
      assetType: r.get('assetType'),
      shareRefId: r.get('sharerRefId'),
      shareDate: r.get('shareDate'),
    };
  });
  return assets;
};
