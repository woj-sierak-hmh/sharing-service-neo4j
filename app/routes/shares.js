import * as shares from '../db/queries/shares.js';

export const createShare = async (ctx, next) => {
  // TODO validation of params

  ctx.log.info('ctx.request.body:', ctx.request.body);
  const recipients = ctx.request.body.recipientRefIds;
  ctx.log.info('recipientRefIds:', recipients);

  try {
    const result = await shares.createShare({
      ...ctx.params,
      recipients: recipients || ['grzuby'],
    });
    ctx.log.info('--------------->', result.summary.statement.parameters);
    ctx.status = 202;
  } catch (err) {
    ctx.log.error(err);
    ctx.status = 500;
    ctx.body = 'Apparently something went wrong...' + err.code;
  }

  // on application exit:
  // driver.close();
};

export const getShares = async ctx => {
  ctx.log.info('params-->', ctx.params);
  try {
    const assets = await shares.getShares(ctx.params);
    ctx.status = 200;
    ctx.body = { assets };
  } catch (err) {
    ctx.log.error(err);
    ctx.status = 500;
    ctx.body = 'Apparently something went wrong...' + err.code;
  }
};
