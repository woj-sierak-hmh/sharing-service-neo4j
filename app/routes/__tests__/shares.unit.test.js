import { createShare, getShares } from '../shares.js';
import * as shares from '../../db/queries/shares.js';

jest.mock('../../db/queries/shares.js');

const createShareCtx = {
  request: {
    body: {
      recipientRefIds: ['teacherA2', 'teacherA3B1'],
    },
  },
  params: {
    tenantRefId: 'districtA',
    sharerRefId: 'teacherA3B1',
    assetType: 'PLAN',
    assetRefId: 'plan2',
  },
  log: {
    info: jest.fn(),
    error: jest.fn(),
  },
  status: null,
  body: null,
};

describe('shares routes', () => {
  beforeAll(() => {
    // ?
  });

  afterEach(() => {
    createShareCtx.log.info.mockClear();
    createShareCtx.log.error.mockClear();
  });

  test('createShare success', async () => {
    shares.createShare = jest.fn().mockResolvedValue({
      summary: {
        statement: {
          parameters: {
            tenantRefId: 'districtA',
            sharerRefId: 'teacherA3B1',
            assetType: 'PLAN',
            assetRefId: 'plan2',
            createdDate: '2019-03-05T21:51:50.609Z',
            recipients: ['teacherA1', 'teacherA2'],
          },
        },
      },
    });

    await createShare(createShareCtx);
    expect(shares.createShare).toHaveBeenCalledWith({
      ...createShareCtx.params,
      recipients: createShareCtx.request.body.recipientRefIds,
    });
    expect(createShareCtx.log.info.mock.calls).toMatchSnapshot();
    expect(createShareCtx.log.error).toHaveBeenCalledTimes(0);
    expect(createShareCtx.status).toBe(202);
  });

  test('createShare db failure', async () => {
    const dbError = { msg: 'some db error', code: 123 };
    shares.createShare = jest.fn().mockRejectedValue(dbError);
    await createShare(createShareCtx);
    expect(shares.createShare).toHaveBeenCalledWith({
      ...createShareCtx.params,
      recipients: createShareCtx.request.body.recipientRefIds,
    });
    expect(createShareCtx.log.info.mock.calls).toMatchSnapshot();
    expect(createShareCtx.log.error.mock.calls[0][0]).toBe(dbError);
    expect(createShareCtx.status).toBe(500);
    expect(createShareCtx.body).toMatchSnapshot();
  });
});
