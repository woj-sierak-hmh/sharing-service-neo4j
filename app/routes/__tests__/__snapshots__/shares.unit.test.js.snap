// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`shares routes createShare db failure 1`] = `
Array [
  Array [
    "ctx.request.body:",
    Object {
      "recipientRefIds": Array [
        "teacherA2",
        "teacherA3B1",
      ],
    },
  ],
  Array [
    "recipientRefIds:",
    Array [
      "teacherA2",
      "teacherA3B1",
    ],
  ],
]
`;

exports[`shares routes createShare db failure 2`] = `"Apparently something went wrong...123"`;

exports[`shares routes createShare success 1`] = `
Array [
  Array [
    "ctx.request.body:",
    Object {
      "recipientRefIds": Array [
        "teacherA2",
        "teacherA3B1",
      ],
    },
  ],
  Array [
    "recipientRefIds:",
    Array [
      "teacherA2",
      "teacherA3B1",
    ],
  ],
  Array [
    "--------------->",
    Object {
      "assetRefId": "plan2",
      "assetType": "PLAN",
      "createdDate": "2019-03-05T21:51:50.609Z",
      "recipients": Array [
        "teacherA1",
        "teacherA2",
      ],
      "sharerRefId": "teacherA3B1",
      "tenantRefId": "districtA",
    },
  ],
]
`;
